### DISCUSSION TITLE
Backbone Border Router

### DISCUSSION THREAD

**Discussion by MichalPoterek** [openthread/openthread#7790] [Q&A]
I would like to test an OpenThread network in two localisations that are far away from each other and there is not possible to establish radio links between these localisations.  I have two BeagleBone BorderRouters with Thread 1.2 equipped with Nordic nRF52840 dongles. Routers are connected to the same LAN network,  one router has IP address 192.168.1.173 and the second 192.168.1.175. Both routers can ping their IPv4 addresses.

![image](https://user-images.githubusercontent.com/67839947/172067478-166386f7-cca0-40c0-a998-0d39e4d6d8aa.png)
![image](https://user-images.githubusercontent.com/67839947/172067516-bacd34a7-1dfc-42a9-a4a8-043965d21aa0.png)

Router 192.168.1.173  has installed Node Red with COAP server which control  devices in the network.  How to set backbone  to see all devices in network and control them?

Below first router configuration

```
root@beaglebone:/home/c# ifconfig
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.1.173  netmask 255.255.255.0  broadcast 192.168.1.255
        inet6 fe80::f73f:9dd6:1065:745c  prefixlen 64  scopeid 0x20<link>
        inet6 fdaa:bbcc:ddee:0:4786:9fa7:999b:be75  prefixlen 64  scopeid 0x0<global>
        inet6 fdaa:bbcc:ddee:0:82f5:b5ff:feb5:dc9c  prefixlen 64  scopeid 0x0<global>
        inet6 2a00:23a8:d86:1201:bfae:56b4:ff4c:9193  prefixlen 64  scopeid 0x0<global>
        inet6 2a00:23a8:d86:1201:82f5:b5ff:feb5:dc9c  prefixlen 64  scopeid 0x0<global>
        ether 80:f5:b5:b5:dc:9c  txqueuelen 1000  (Ethernet)
        RX packets 201205  bytes 23093623 (22.0 MiB)
        RX errors 0  dropped 454  overruns 0  frame 0
        TX packets 159245  bytes 77247812 (73.6 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
        device interrupt 54

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 33  bytes 6119 (5.9 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 33  bytes 6119 (5.9 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

nat64: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 1500
        inet 192.168.255.1  netmask 255.255.255.255  destination 192.168.255.1
        inet6 fe80::761c:8831:4b7c:5f94  prefixlen 64  scopeid 0x20<link>
        inet6 fdaa:bb:1::2  prefixlen 128  scopeid 0x0<global>
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 500  (UNSPEC)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 4  bytes 304 (304.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

wpan0: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 1280
        inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc35  prefixlen 64  scopeid 0x0<global>
        inet6 fe80::9cd5:5117:394b:eef  prefixlen 64  scopeid 0x20<link>
        inet6 fd3f:6880:77d1:6e92:c0ca:41cb:221:5326  prefixlen 64  scopeid 0x0<global>
        inet6 fd8f:c605:8837:1:1257:f9e6:99da:1e06  prefixlen 64  scopeid 0x0<global>
        inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc11  prefixlen 64  scopeid 0x0<global>
        inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc00  prefixlen 64  scopeid 0x0<global>
        inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc38  prefixlen 64  scopeid 0x0<global>
        inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc33  prefixlen 64  scopeid 0x0<global>
        inet6 fd3f:6880:77d1:6e92:0:ff:fe00:1c00  prefixlen 64  scopeid 0x0<global>
        inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc10  prefixlen 64  scopeid 0x0<global>
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 500  (UNSPEC)
        RX packets 13242  bytes 2332558 (2.2 MiB)
        RX errors 0  dropped 1  overruns 0  frame 0
        TX packets 5711  bytes 643456 (628.3 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
ipaddr
fd3f:6880:77d1:6e92:0:ff:fe00:fc33
fd3f:6880:77d1:6e92:0:ff:fe00:fc35
fd3f:6880:77d1:6e92:0:ff:fe00:fc11
fd8f:c605:8837:1:1257:f9e6:99da:1e06
fd3f:6880:77d1:6e92:0:ff:fe00:fc10
fd3f:6880:77d1:6e92:0:ff:fe00:fc38
fd3f:6880:77d1:6e92:0:ff:fe00:fc00
fd3f:6880:77d1:6e92:0:ff:fe00:1c00
fd3f:6880:77d1:6e92:c0ca:41cb:221:5326
fe80:0:0:0:9cd5:5117:394b:eef
Done
> ipmaddr
ff02:0:0:0:0:0:0:16
ff33:40:fd3f:6880:77d1:6e92:0:1
ff32:40:fd3f:6880:77d1:6e92:0:1
ff02:0:0:0:0:0:0:2
ff03:0:0:0:0:0:0:2
ff02:0:0:0:0:0:0:1
ff03:0:0:0:0:0:0:1
ff03:0:0:0:0:0:0:fc
Done
> route
fdaa:bbcc:ddee:0::/64 s med 1c00
2a00:23a8:d86:1201::/64 s med 1c00
fdbd:6359:d328:1::/64 s med 1c00
Done
> bbr config
seqno:    125
delay:    1200 secs
timeout:  3600 secs
Done
> bbr state
Primary
Done
```

And the second configuration

```
root@beaglebone2:/home/c# ifconfig
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.1.175  netmask 255.255.255.0  broadcast 192.168.1.255
        inet6 2a00:23a8:d86:1201:34de:3663:b7be:51e5  prefixlen 64  scopeid 0x0<global>
        inet6 fdaa:bbcc:ddee:0:761b:24d:3863:c05e  prefixlen 64  scopeid 0x0<global>
        inet6 fe80::b52c:9b4d:348a:80ca  prefixlen 64  scopeid 0x20<link>
        ether 24:76:25:40:e5:29  txqueuelen 1000  (Ethernet)
        RX packets 121198  bytes 20561115 (19.6 MiB)
        RX errors 0  dropped 500  overruns 0  frame 0
        TX packets 10586  bytes 8177522 (7.7 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
        device interrupt 54

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 100  bytes 19616 (19.1 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 100  bytes 19616 (19.1 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

nat64: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 1500
        inet 192.168.255.1  netmask 255.255.255.255  destination 192.168.255.1
        inet6 fe80::47e8:4bc4:1a3d:fa6a  prefixlen 64  scopeid 0x20<link>
        inet6 fdaa:bb:1::2  prefixlen 128  scopeid 0x0<global>
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 500  (UNSPEC)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 4  bytes 304 (304.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

wpan0: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 1280
        inet6 fd55:6d5f:5832:382f:0:ff:fe00:fc10  prefixlen 64  scopeid 0x0<global>
        inet6 fd55:6d5f:5832:382f:0:ff:fe00:fc11  prefixlen 64  scopeid 0x0<global>
        inet6 fd55:6d5f:5832:382f:915a:c130:6f80:fc1e  prefixlen 64  scopeid 0x0<global>
        inet6 fdbd:6359:d328:1:6264:56da:2296:2415  prefixlen 64  scopeid 0x0<global>
        inet6 fd55:6d5f:5832:382f:0:ff:fe00:fc00  prefixlen 64  scopeid 0x0<global>
        inet6 fd55:6d5f:5832:382f:0:ff:fe00:ac00  prefixlen 64  scopeid 0x0<global>
        inet6 fe80::3416:ec4b:c1dd:b114  prefixlen 64  scopeid 0x20<link>
        inet6 fd55:6d5f:5832:382f:0:ff:fe00:fc38  prefixlen 64  scopeid 0x0<global>
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 500  (UNSPEC)
        RX packets 15  bytes 1560 (1.5 KiB)
        RX errors 0  dropped 2  overruns 0  frame 0
        TX packets 13  bytes 2548 (2.4 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
ipaddr
fd55:6d5f:5832:382f:0:ff:fe00:fc10
fd55:6d5f:5832:382f:0:ff:fe00:fc38
fdbd:6359:d328:1:6264:56da:2296:2415
fd55:6d5f:5832:382f:0:ff:fe00:fc11
fd55:6d5f:5832:382f:0:ff:fe00:fc00
fd55:6d5f:5832:382f:0:ff:fe00:ac00
fd55:6d5f:5832:382f:915a:c130:6f80:fc1e
fe80:0:0:0:3416:ec4b:c1dd:b114
Done
> ipmaddr
ff02:0:0:0:0:0:0:16
ff33:40:fd55:6d5f:5832:382f:0:1
ff32:40:fd55:6d5f:5832:382f:0:1
ff02:0:0:0:0:0:0:2
ff03:0:0:0:0:0:0:2
ff02:0:0:0:0:0:0:1
ff03:0:0:0:0:0:0:1
ff03:0:0:0:0:0:0:fc
Done
> route
fdaa:bbcc:ddee:0::/64 s med ac00
2a00:23a8:d86:1201::/64 s med ac00
fd8f:c605:8837:1::/64 s med ac00
Done
> bbr state
Primary
Done
> bbr config
seqno:    20
delay:    30 secs
timeout:  3600 secs
Done
```

**Comment by jwhui** [comment]
What ot-br-posix git commit are you using?

If you are using the latest `ot-br-posix` main branch, the Border Routers should automatically configure:
1. Routable IPv6 prefixes for the different Thread networks and the infrastructure (i.e. Ethernet) link.
2. Routes to the the routable IPv6 prefix of each Thread network.

The output of `netdata show` can help indicate what prefixes and routes are configured on the Thread network.

**Reply by MichalPoterek** [reply]
First beaglebone Border Router with IP address 192.168.1.173 

```
root@beaglebone:/home/c# ot-ctl
> netdata show
Prefixes:
fd8f:c605:8837:1::/64 paos med 1c00
Routes:
fdbd:6359:d328:1::/64 s med 1c00
fdaa:bbcc:ddee:0::/64 s med 1c00
2a00:23a8:d86:1201::/64 s med 1c00
Services:
44970 01 0004b000000e10 s 1c00
44970 5d fd3f688077d16e92c0ca41cb02215326d11f s 1c00
Done

```

Sedond beaglebone Border Router with ip address 192.168.1.175 

```
 netdata show
Prefixes:
fdbd:6359:d328:1::/64 paos med ac00
Routes:
fdaa:bbcc:ddee:0::/64 s med ac00
2a00:23a8:d86:1201::/64 s med ac00
fd8f:c605:8837:1::/64 s med ac00
Services:
44970 01 16001e00000e10 s ac00
44970 5d fd556d5f5832382f915ac1306f80fc1ed11f s ac00
Done

```


I'm using the following commit for both routers

```
c@beaglebone2:/home/debian/ot-br-posix$ git show
commit a4ec0d4716c87a4a6c4206d0fca27a74cf69e39e (HEAD -> main, origin/main, origin/HEAD)
Author: Abtin Keshavarzian <abtink@google.com>
Date:   Tue May 3 19:14:07 2022 -0700

    [adv-proxy] treat `NOT_FOUND` as success when unpublishing host/service (#1355)

    This commit updates `AdvertisingProxy` so that when it is unpublishing
    a host or service entry on the underlying `Mdns::Publisher` if the
    returned error is `OTBR_ERROR_NOT_FOUND` the error is treated as
    success and reported as such to OT core `Srp::Server`. This ensures
    that an SRP update request to remove a not previously registered
    host/service is not rejected by the platform and therefore by the
    `ot::Srp:Server`.

diff --git a/src/sdp_proxy/advertising_proxy.cpp b/src/sdp_proxy/advertising_proxy.cpp
index 9b8ddfe42c..336eeeb8a4 100644
--- a/src/sdp_proxy/advertising_proxy.cpp
+++ b/src/sdp_proxy/advertising_proxy.cpp
@@ -261,6 +261,8 @@ otbrError AdvertisingProxy::PublishHostAndItsServices(const otSrpServerHost *aHo
             otbrLogDebug("Unpublish SRP service '%s'", fullServiceName.c_str());
             mPublisher.UnpublishService(
                 serviceName, serviceType, [this, hasUpdate, updateId, fullServiceName](otbrError aError) {
+                    // Treat `NOT_FOUND` as success when unpublishing service
+                    aError = (aError == OTBR_ERROR_NOT_FOUND) ? OTBR_ERROR_NONE : aError;
                     otbrLogResult(aError, "Handle unpublish SRP service '%s'", fullServiceName.c_str());
                     if (hasUpdate)
                     {
@@ -291,6 +293,8 @@ otbrError AdvertisingProxy::PublishHostAndItsServices(const otSrpServerHost *aHo
     {
         otbrLogDebug("Unpublish SRP host '%s'", fullHostName.c_str());
         mPublisher.UnpublishHost(hostName, [this, hasUpdate, updateId, fullHostName](otbrError aError) {
+            // Treat `NOT_FOUND` as success when unpublishing host.
+            aError = (aError == OTBR_ERROR_NOT_FOUND) ? OTBR_ERROR_NONE : aError;
             otbrLogResult(aError, "Handle unpublish SRP host '%s'", fullHostName.c_str());
             if (hasUpdate)
             {

root@beaglebone2:/home/debian/ot-br-posix# git status
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean

```

**Reply by jwhui** [reply]
It appears that both of your Thread networks have:
1. A routable IPv6 prefix (fd8f:c605:8837:1::/64 and fdbd:6359:d328:1::/64) configured on the Thread network.
2. A route to the other Thread network's IPv6 prefix (fd8f:c605:8837:1::/64 and fdbd:6359:d328:1::/64).
3. A routable IPv6 prefix configured on the infrastructure link (fdaa:bbcc:ddee:0::/64).

This should be sufficient for communicating with devices in both Thread networks. You should be able to ping Thread devices from a host on the infrastructure network or other Thread devices (including the border routers).

**Comment by MichalPoterek** [comment]
Thank you Jonathan for your help.  I cannot manage to make my routers work as I want.
So I have two BR routers one called beaglebone and the second beaglebone2.  Both BR are connected with Ethernet cable through not manageable simple switch. On beaglebone I have installed Node Red with Coap which control end devices. All devices connected to beaglebone can be controlled and I can ping them. Devices connected to beaglebone2 cannot reach beaglebone Node Red Coap service.

On beaglebone I set fix IP address with following script.

#!/bin/bash
echo "Setting Fix Address to wpan0 interface"
ot-ctl ipaddr add fd3f:6880:77d1:6e92:0:ff:fe00:fc33
ot-ctl thread stop
ot-ctl thread start
sleep 2s
ot-ctl commissioner start
sleep 5s
ot-ctl commissioner joiner add * J01NME
ip -f inet6 addr add fd3f:6880:77d1:6e92:0:ff:fe00:fc33/64 dev wpan0

To be able to ping this adress from beaglebone2 I  issued following commands on beaglebone2
route -6 add fd3f:6880:77d1:6e92::/64 metric 65535 dev wpan0
route -6 add fd3f:6880:77d1:6e92::/64 gw fe80::82f5:b5ff:feb5:dc9c dev eth0


and following on beaglebone
route -6 add fd3f:6880:77d1:6e92:0:ff:fe00:fc33/128 dev wpan0

then I was able to ping  beaglebone fix ipv6 wpan0 addres fd3f:6880:77d1:6e92:0:ff:fe00:fc33 from bash on beaglebone2, but I cannot ping this address by using ot-ctl ping  fd3f:6880:77d1:6e92:0:ff:fe00:fc33

Below picture showing my problem
![image](https://user-images.githubusercontent.com/67839947/172699202-52eacc32-8be6-49b8-b159-3d0e40e85448.png)

And on the below picture, you can see I can ping from beaglebone2 bash end device connected to beaglebone over Thread radio but I cannot do it by using ot-ctl ping on beaglebone2.
![image](https://user-images.githubusercontent.com/67839947/172700531-383a5193-c19f-4ecd-b9f3-d7c3245f438a.png)




I will be appreciated If you give me hints how to make possible to ping from beaglebone2  (ot-ctl ping)   fix address assigned to beaglebone wpan0 interface fd3f:6880:77d1:6e92:0:ff:fe00:fc33.  If I achieve it I should be able to control devices connected to beaglebone2 from   beaglebone Node Red Coap service.

**Reply by jwhui** [reply]
Some comments:
1. You do not need to add your own IPv6 address. You can use the one automatically configured for you (fd3f:6880:77d1:6e92:c0ca:41cb:221:5326).
2. You should not need to add your own route. A route should already be configured. You can view the Linux routing table with `route -A inet6`.
3. Use `tcpdump -i wpan0` to observe IPv6 packets going in and coming out of the `wpan0` interface.

**Comment by MichalPoterek** [comment]
Does BR need radvd operating? In my case it's  not operating.
root@beaglebone:/home/c# ps -A|grep radvd
root@beaglebone:/home/c#

**Reply by jwhui** [reply]
No, OpenThread sends its own Router Advertisement messages. You can observe this using `tcpdump` on the infrastructure interface (e.g. eth0 or wlan0).

**Comment by MichalPoterek** [comment]
I learn quite a lot from your answers Jonatan.  What really wonders is that both BR show bbr state Primary. I guess one should be Primary and the others Secondary/Slave? (both are connected to the same simple not managed ethernet switch)

**Reply by jwhui** [reply]
For what you are trying to achieve, I do not think you need to have Thread 1.2 Backbone Router. All you need is IPv6 reachability between different hosts on your infrastructure network. It is fine that both your border routers are showing up as primary.

**Reply by MichalPoterek** [reply]
I would like to understand OpenThread routing well. Below is my router configuration. Please correct me if I understand something wrong.
Addreses beggining with fd are so called Mesh Local Addresses. These are assigned to my wpan0 interface
```
inet6 fd3f:6880:77d1:6e92:c0ca:41cb:221:5326  prefixlen 64  scopeid 0x0<global>
**inet6 fd8f:c605:8837:1:1257:f9e6:99da:1e06  prefixlen 64  scopeid 0x0<global>**
inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc11  prefixlen 64  scopeid 0x0<global>
inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc00  prefixlen 64  scopeid 0x0<global>
inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc37  prefixlen 64  scopeid 0x0<global>
inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc38  prefixlen 64  scopeid 0x0<global>
inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc33  prefixlen 128  scopeid 0x0<global>
inet6 fd3f:6880:77d1:6e92:0:ff:fe00:6000  prefixlen 64  scopeid 0x0<global>
inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc10  prefixlen 64  scopeid 0x0<global>
```

There are 3 types one is Mesh-Local EID (ML-EID) , Routing Locator (RLOC), Anycast Locator (ALOC)
**Does ML-EID addresses are assigned only to Router Eligible End Devises, Full End Devises, Sleepy End Devices, Minimal End Devices and are not assigned to eny other devices? Is otIp6GetUnicastAddresses function used like below is correct way to get Routing Locator Interface IP address?**


```
const otNetifAddress *ip_address;
                                otInstance *instance = OtInstance_get();
                                
                                ip_address = otIp6GetUnicastAddresses(instance);
                                for (const otNetifAddress *addr = ip_address; addr; addr = addr->mNext)
                                 {

                                     if(addr->mRloc==1)
                                     {
                                             ToString(addr, ThisDeviceIP );

                                     };
                                 };
   
```                             

**How to get  ML-EID address (I cannot see how to distinguish ML-EID address in otNetifAddrress structure)?**

```
typedef struct otNetifAddress
{
    otIp6Address           mAddress;                ///< The IPv6 unicast address.
    uint8_t                mPrefixLength;           ///< The Prefix length (in bits).
    bool                   mPreferred : 1;          ///< TRUE if the address is preferred, FALSE otherwise.
    bool                   mValid : 1;              ///< TRUE if the address is valid, FALSE otherwise.
    bool                   mScopeOverrideValid : 1; ///< TRUE if the mScopeOverride value is valid, FALSE otherwise.
    unsigned int           mScopeOverride : 4;      ///< The IPv6 scope of this address.
    bool                   mRloc : 1;               ///< TRUE if the address is an RLOC, FALSE otherwise.
    struct otNetifAddress *mNext;                   ///< A pointer to the next network interface address.
} otNetifAddress;

```
**Does ML-EID will be not present/assigned  if device is router type?** 

How to get  ML-ALOC address (I cannot see how to distinguish ML-EID address in otNetifAddrress structure)? Will ML-ALOC be not assigned if the device hasn't any of the following services: leader, dhcp agent, commissioner, neighbor discovery agent?


And very important to me question. I have 2 BR based on clonned sd cards. The same OpenThread settings like OpenThread network name,  and master key etc.

**Why addresses assigned to wpan0  on BR1  not answer when pinging them from BR2?**
inet6 fd3f:6880:77d1:6e92:c0ca:41cb:221:5326  prefixlen 64  scopeid 0x0<global>
inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc11  prefixlen 64  scopeid 0x0<global>
inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc00  prefixlen 64  scopeid 0x0<global>
inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc37  prefixlen 64  scopeid 0x0<global>
inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc38  prefixlen 64  scopeid 0x0<global>
inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc33  prefixlen 128  scopeid 0x0<global>
inet6 fd3f:6880:77d1:6e92:0:ff:fe00:6000  prefixlen 64  scopeid 0x0<global>
inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc10  prefixlen 64  scopeid 0x0<global>

**but address **inet6 fd8f:c605:8837:1:1257:f9e6:99da:1e06  prefixlen 64  scopeid 0x0<global>**    which is also assigned to wpan0 on BR1 is replaying ping?** It looks like router advertise route to this address and other routers know the route to it. **Why routes to other addresses assigned to wpan0 on BR1 are not advertised?**

**How to know to which address assigned to wpan0 routes are advertised?**

In the network we are designing we using only Full Tred Devices powered from grid. Some of them are routers and some of them are End devices. **Which address should I use to send Coap message to it from BR1  (Shall I use above function and check if address is RLOC type and use this address, if not how to obtain this address)?** 

Currently all devices once connected to OpenThread network are sending Coap messages with thir IP addresses do Node Red installed on BR1. I'm currently using following function which getting IP address which later is included in Coap payload

```
const otNetifAddress *ip_address;
                                otInstance *instance = OtInstance_get();
                                ip_address = otIp6GetUnicastAddresses(instance);

                                ToString(ip_address, tmpThisDeviceIP );
```

I have taken wpan0 ip address fd3f:6880:77d1:6e92:0:ff:fe00:fc33 as a destination point for coap messages sended to Node Red from end devices/routers. Unfortunatelly by doing that routers/end devices connected to BR2  are not seen in Node red. 
When I'm using this BR1 wpan0 ip address fd8f:c605:8837:1:1257:f9e6:99da:1e06 coap messages reach Node Read, but I cannot  control end devices/routers when I sending to them coap messages with use ip addresses which was sended in coap payload (obtained by above function ip_address = otIp6GetUnicastAddresses(instance))


BR1 router configuration

```
eth0: flags=-28605<UP,BROADCAST,RUNNING,MULTICAST,DYNAMIC>  mtu 1500
        inet 192.168.1.173  netmask 255.255.255.0  broadcast 192.168.1.255
        inet6 fdaa:bbcc:ddee:0:82f5:b5ff:feb5:dc9c  prefixlen 64  scopeid 0x0<global>
        inet6 fe80::82f5:b5ff:feb5:dc9c  prefixlen 64  scopeid 0x20<link>
        inet6 2a00:23a8:d86:1201:82f5:b5ff:feb5:dc9c  prefixlen 64  scopeid 0x0<global>
        ether 80:f5:b5:b5:dc:9c  txqueuelen 1000  (Ethernet)
        RX packets 425836  bytes 111872549 (106.6 MiB)
        RX errors 0  dropped 1549  overruns 0  frame 0
        TX packets 328568  bytes 283097759 (269.9 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
        device interrupt 54

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 27  bytes 4036 (3.9 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 27  bytes 4036 (3.9 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

nat64: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 1500
        inet 192.168.255.1  netmask 255.255.255.255  destination 192.168.255.1
        inet6 fe80::9553:50c4:fbc0:9f5e  prefixlen 64  scopeid 0x20<link>
        inet6 fdaa:bb:1::2  prefixlen 128  scopeid 0x0<global>
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 500  (UNSPEC)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 4  bytes 304 (304.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

wpan0: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 1280
        inet6 fe80::9cd5:5117:394b:eef  prefixlen 64  scopeid 0x20<link>
        inet6 fd3f:6880:77d1:6e92:c0ca:41cb:221:5326  prefixlen 64  scopeid 0x0<global>
        inet6 fd8f:c605:8837:1:1257:f9e6:99da:1e06  prefixlen 64  scopeid 0x0<global>
        inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc11  prefixlen 64  scopeid 0x0<global>
        inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc00  prefixlen 64  scopeid 0x0<global>
        inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc37  prefixlen 64  scopeid 0x0<global>
        inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc38  prefixlen 64  scopeid 0x0<global>
        inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc33  prefixlen 128  scopeid 0x0<global>
        inet6 fd3f:6880:77d1:6e92:0:ff:fe00:6000  prefixlen 64  scopeid 0x0<global>
        inet6 fd3f:6880:77d1:6e92:0:ff:fe00:fc10  prefixlen 64  scopeid 0x0<global>
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 500  (UNSPEC)
        RX packets 12049  bytes 3295338 (3.1 MiB)
        RX errors 0  dropped 1  overruns 0  frame 0
        TX packets 501  bytes 58600 (57.2 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

root@beaglebone:/home/c# ot-ctl prefix
fd8f:c605:8837:1::/64 paos med 6000
Done
root@beaglebone:/home/c# ot-ctl prefix meshlocal
fd3f:6880:77d1:6e92::/64
Done
Done
root@beaglebone:/home/c# ot-ctl
netdata show
Prefixes:
fd8f:c605:8837:1::/64 paos med 6000
Routes:
fdaa:bbcc:ddee:0::/64 s med 6000
2a00:23a8:d86:1201::/64 s med 6000
fd6b:a9a7:d4b4:1::/64 s med 6000
Services:
44970 01 4e04b000000e10 s 6000
44970 5d fd3f688077d16e92c0ca41cb02215326d11f s 6000
Done
> eidcache
fd8f:c605:8837:1:cc6a:c4b0:7b04:189b 1400 cache canEvict=1
fd8f:c605:8837:1:fdc2:b328:8988:3d21 0c00 cache canEvict=1
fd8f:c605:8837:1:88cd:c778:5b14:50cf 3000 cache canEvict=1
fd8f:c605:8837:1:c5ad:f4b0:db26:c034 fffe retry canEvict=1 timeout=0 retryDelay=120
fd8f:c605:8837:1:cdfb:73f0:ba1f:70d8 fffe retry canEvict=1 timeout=0 retryDelay=120
fd8f:c605:8837:1:bb37:2682:ef9e:71c8 fffe retry canEvict=1 timeout=0 retryDelay=120
fd8f:c605:8837:1:6f09:137f:5300:8a34 fffe retry canEvict=1 timeout=0 retryDelay=120
fd8f:c605:8837:1:a47b:305b:ff3e:5846 fffe retry canEvict=1 timeout=0 retryDelay=120
fd8f:c605:8837:1:d5c4:6212:e201:2fac fffe retry canEvict=1 timeout=0 retryDelay=120
fd8f:c605:8837:1:a569:58c2:4176:2519 fffe retry canEvict=1 timeout=0 retryDelay=120
fd8f:c605:8837:1:1afe:364b:a698:527c fffe retry canEvict=1 timeout=0 retryDelay=120
Done

**root@beaglebone:/home/c# ping6 -I wpan0 ff03::1
PING ff03::1(ff03::1) from fd8f:c605:8837:1:1257:f9e6:99da:1e06 wpan0: 56 data bytes
64 bytes from fd8f:c605:8837:1:1257:f9e6:99da:1e06: icmp_seq=1 ttl=64 time=0.408 ms
64 bytes from fd8f:c605:8837:1:fdc2:b328:8988:3d21: icmp_seq=1 ttl=64 time=27.7 ms
64 bytes from fd8f:c605:8837:1:88cd:c778:5b14:50cf: icmp_seq=1 ttl=64 time=38.0 ms
64 bytes from fd8f:c605:8837:1:cc6a:c4b0:7b04:189b: icmp_seq=1 ttl=64 time=54.7 ms**
```

**Reply by jwhui** [reply]
> Does ML-EID addresses are assigned only to Router Eligible End Devises, Full End Devises, Sleepy End Devices, Minimal End Devices and are not assigned to eny other devices?

ML-EID is assigned to all Thread interfaces, regardless of role.

> Does ML-EID will be not present/assigned if device is router type?

ML-EID is assigned to all Thread interfaces, regardless of role.

> Why addresses assigned to wpan0 on BR1 not answer when pinging them from BR2?

Mesh Local Addresses are scoped to a given Thread Partition. They are not routable outside the Thread partition.

In your example:
- ` fd3f:6880:77d1:6e92::/64` is a mesh-local prefix.
- `fd8f:c605:8837:1::/64` is a ULA prefix, which has global scope and is routable outside the Thread partition.

It is expected that the BR will only advertise a route to the ULA prefix on the infrastructure network. You should see `fd8f:c605:8837:1::/64` in Router Advertisement messages, but not `fd3f:6880:77d1:6e92::/64`.

> How to know to which address assigned to wpan0 routes are advertised?

A BR advertises routes to all IPv6 prefixes that are neither mesh-local nor link-local in Router Advertisement messages on the infrastructure network.

> Which address should I use to send Coap message to it from BR1 (Shall I use above function and check if address is RLOC type and use this address, if not how to obtain this address)?

You should use the ULA address (`fd8f:c605:8837:1::/64`) since that is neither mesh-local nor link-local.
